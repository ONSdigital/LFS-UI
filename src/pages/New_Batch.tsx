import React, {Component, ChangeEvent} from "react";
import {ONSSelect} from "../components/ONSSelect";
import {ONSSubmitButton} from "../components/ONSSubmitButton"
import {ONSButton} from "../components/ONSButton";
import {runs, years, quarters, months } from "../utilities/input_fields";

interface State{
    runType: String;
    year: String;
    period: String;
    submitError?: boolean;
}

export class New_Batch extends Component <{}, State> {
    displayName = New_Batch.name
    constructor(props: any) {
        super(props);
        this.state = {
            runType: "",
            year: "",
            period: "",
            submitError: false
        };
    }
    error = false;

    handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        //setting up for the panel error, if submit === true then there is at least one empty field and submit has been clicked, else redirect to manage run
        if(this.state.runType === "" || this.state.year === "" || (this.state.runType !== "yearly" && this.state.period === "")) this.error = true
        else window.location.href = "/Manage_Run"

        if(this.state.runType !== "yearly") this.setState({period: ""})
        
        this.setState({submitError: this.error})
        
        // printing to console the info to be sent to api on submit
        console.log("subbbbmmmiiiiittttt")
        console.log("run type :" + this.state.runType)
        console.log("run type :" + this.state.year)
        if(this.state.runType !== "yearly") console.log("run type :" + this.state.period)

    };

    handleRunTypeChange = (e: ChangeEvent<HTMLSelectElement>) =>{
        this.setState({runType: e.target.value})
    }
    
    handleYearChange = (e: ChangeEvent<HTMLSelectElement>) =>{
        this.setState({year: e.target.value})
    }

    handlePeriodChange = (e: ChangeEvent<HTMLSelectElement>) =>{
        this.setState({period: e.target.value})
    }

    render() {
        return (
            <div className="container">
                {/* this is for the empty inputs, the  ifs are pure jank but it works */}
                {(() => {if (this.state.submitError === true) {
                    return(
                        <div>
                            <div className="panel panel--error">
                                <div className="panel__header">
                                    <h1 className="panel__title u-fs-r--b">This page has errors</h1>
                                </div>
                                <div className="panel__body">
                                    <ul className="list list--bare">
                                        {(() => {if (this.state.runType === ""){
                                            return(<li className="list__item ">
                                                <p className="list__link js-inpagelink">
                                                Please select a Run Type.
                                                </p>
                                            </li>
                                            )}
                                        })()}
                                        {(() => {if (this.state.year === ""){
                                            return(<li className="list__item ">
                                                <p className="list__link js-inpagelink">
                                                Please select a Year.
                                                </p>
                                            </li>
                                            )}
                                        })()}
                                        {(() => {if ((this.state.runType !== "yearly" && this.state.runType !== "") && this.state.period === ""){
                                            return(<li className="list__item ">
                                                <p className="list__link js-inpagelink">
                                                Please select a Period.
                                                </p>
                                            </li>
                                            )}
                                        })()}
                                    </ul>
                                </div>
                            </div>
                            <div>
                                <br/>
                            </div>
                        </div>
                    )}
                })()}

                <form onSubmit={this.handleSubmit}>
                    <div style={{maxWidth: "351px"}}>
                        <ONSSelect label="Run Type" value="run" options={runs} onChange={this.handleRunTypeChange}/>
                        <ONSSelect label="Year" value="year" options={years()} onChange={this.handleYearChange} />
                        {(() => {if (this.state.runType === "monthly") {
                            return(
                                <ONSSelect label="Period" value="period" options={months()} onChange={this.handlePeriodChange} />
                            )
                        }if(this.state.runType === "quarterly"){
                            return(
                                <ONSSelect label="Period" value="period" options={quarters} onChange={this.handlePeriodChange} />
                            )
                        }
                        })()}
                        <ONSSubmitButton label="Submit" primary={true} small={false}/>
                    </div>
                </form>
            </div>
        )
    }
}
