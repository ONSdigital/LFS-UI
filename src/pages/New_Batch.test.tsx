import React from "react"

import { New_Batch } from "./New_Batch";

import {cleanup, fireEvent, render, screen} from "@testing-library/react";

import {BrowserRouter} from "react-router-dom";
import {act} from "react-dom/test-utils";

import fetch from "../tests/setup/__mocks__/fetch.js";
import {default as flushPromises} from "../tests/util/flushPromises";
import { ReactWrapper } from "enzyme";

// @ts-ignore
global.fetch = fetch;

describe("New batch creation - ", () => {
    
    afterEach(() => {
        cleanup();
    });
    
    function wrapper(render: any) {
        return render(
            <BrowserRouter>
                <New_Batch/>
            </BrowserRouter>
        );
    }
    
    it("renders the page", async () => {
        expect(wrapper(render)).toMatchSnapshot();
    });
    
    test("The page redirects successfully on a happy yearly path", async () => {
        wrapper(render)

        // Selecting from drop downs
        await fireEvent.change(screen.getByLabelText(/Batch Type/i), {
            target: {value: "yearly"}
        }); 

        await fireEvent.change(screen.getByLabelText(/Year/i), {
            target: {value: "2015"}
        }); 

        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await act(async () => {
            await flushPromises();
        });
        
        // expect(screen.getByText(/this page has errors/i))
    })

    test("The page redirects successfully on a quarterly path", async () => {
        wrapper(render)

        // Selecting from drop downs
        await fireEvent.change(screen.getByLabelText(/Batch Type/i), {
            target: {value: "quarterly"}
        }); 

        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await fireEvent.change(screen.getByLabelText(/Year/i), {
            target: {value: "2015"}
        }); 

        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await fireEvent.change(screen.getByLabelText(/Period/i), {
            target: {value: "Q1"}
        }); 

        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await act(async () => {
            await flushPromises();
        });
        
        // expect(screen.getByText(/this page has errors/i))
    })

    test("The page redirects successfully on a monthly path", async () => {
        wrapper(render)

        // Selecting from drop downs
        await fireEvent.change(screen.getByLabelText(/Batch Type/i), {
            target: {value: "monthly"}
        }); 
        
        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await fireEvent.change(screen.getByLabelText(/Year/i), {
            target: {value: "2015"}
        }); 
        
        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await fireEvent.change(screen.getByLabelText(/Period/i), {
            target: {value: "1"}
        }); 

        fireEvent.submit(screen.getByTestId(/New-Batch-Form/i))

        await act(async () => {
            await flushPromises();
        });
        
        // expect(screen.getByText(/this page has errors/i))
    })
    
})