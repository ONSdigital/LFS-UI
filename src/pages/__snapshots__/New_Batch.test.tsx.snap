// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New batch creation -  renders the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <nav
          aria-label="Breadcrumb"
          class="breadcrumb"
          data-testid="breadcrumbs"
        >
          <ol
            class="breadcrumb__items"
          >
            <li
              class="breadcrumb__item u-fs-s"
              id="breadcrumb-1"
            >
              <a
                class="breadcrumb__link"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="breadcrumb__item u-fs-s breadcrumb__item--current"
              data-testid="breadcrumb-current"
              id="breadcrumb-2"
            >
              Create New Batch
            </li>
          </ol>
        </nav>
        <form
          data-testid="New-Batch-Form"
        >
          <div
            style="max-width: 351px;"
          >
            <p
              class="field"
            >
              <label
                class="label"
                for="batch"
              >
                Batch Type
                 
              </label>
              <select
                class="input input--select"
                id="batch"
                name="select"
              >
                <option
                  data-testid="select-batch"
                  disabled=""
                  selected=""
                  value=""
                >
                  Select an option
                </option>
                <option
                  data-testid="option-batch-monthly"
                  value="monthly"
                >
                  Monthly
                </option>
                <option
                  data-testid="option-batch-quarterly"
                  value="quarterly"
                >
                  Quarterly
                </option>
                <option
                  data-testid="option-batch-yearly"
                  value="yearly"
                >
                  Yearly
                </option>
              </select>
            </p>
            <p
              class="field"
            >
              <label
                class="label"
                for="year"
              >
                Year
                 
              </label>
              <select
                class="input input--select"
                id="year"
                name="select"
              >
                <option
                  data-testid="select-year"
                  disabled=""
                  selected=""
                  value=""
                >
                  Select an option
                </option>
                <option
                  data-testid="option-year-2015"
                  value="2015"
                >
                  2015
                </option>
                <option
                  data-testid="option-year-2014"
                  value="2014"
                >
                  2014
                </option>
                <option
                  data-testid="option-year-2013"
                  value="2013"
                >
                  2013
                </option>
                <option
                  data-testid="option-year-2012"
                  value="2012"
                >
                  2012
                </option>
                <option
                  data-testid="option-year-2011"
                  value="2011"
                >
                  2011
                </option>
                <option
                  data-testid="option-year-2010"
                  value="2010"
                >
                  2010
                </option>
                <option
                  data-testid="option-year-2009"
                  value="2009"
                >
                  2009
                </option>
                <option
                  data-testid="option-year-2008"
                  value="2008"
                >
                  2008
                </option>
                <option
                  data-testid="option-year-2007"
                  value="2007"
                >
                  2007
                </option>
                <option
                  data-testid="option-year-2006"
                  value="2006"
                >
                  2006
                </option>
              </select>
            </p>
            <button
              class="field btn "
              data-testid="submit-button"
              id="newBatchSubmit"
              type="submit"
            >
              <span
                class="btn__inner"
              >
                Submit
              </span>
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <nav
        aria-label="Breadcrumb"
        class="breadcrumb"
        data-testid="breadcrumbs"
      >
        <ol
          class="breadcrumb__items"
        >
          <li
            class="breadcrumb__item u-fs-s"
            id="breadcrumb-1"
          >
            <a
              class="breadcrumb__link"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="breadcrumb__item u-fs-s breadcrumb__item--current"
            data-testid="breadcrumb-current"
            id="breadcrumb-2"
          >
            Create New Batch
          </li>
        </ol>
      </nav>
      <form
        data-testid="New-Batch-Form"
      >
        <div
          style="max-width: 351px;"
        >
          <p
            class="field"
          >
            <label
              class="label"
              for="batch"
            >
              Batch Type
               
            </label>
            <select
              class="input input--select"
              id="batch"
              name="select"
            >
              <option
                data-testid="select-batch"
                disabled=""
                selected=""
                value=""
              >
                Select an option
              </option>
              <option
                data-testid="option-batch-monthly"
                value="monthly"
              >
                Monthly
              </option>
              <option
                data-testid="option-batch-quarterly"
                value="quarterly"
              >
                Quarterly
              </option>
              <option
                data-testid="option-batch-yearly"
                value="yearly"
              >
                Yearly
              </option>
            </select>
          </p>
          <p
            class="field"
          >
            <label
              class="label"
              for="year"
            >
              Year
               
            </label>
            <select
              class="input input--select"
              id="year"
              name="select"
            >
              <option
                data-testid="select-year"
                disabled=""
                selected=""
                value=""
              >
                Select an option
              </option>
              <option
                data-testid="option-year-2015"
                value="2015"
              >
                2015
              </option>
              <option
                data-testid="option-year-2014"
                value="2014"
              >
                2014
              </option>
              <option
                data-testid="option-year-2013"
                value="2013"
              >
                2013
              </option>
              <option
                data-testid="option-year-2012"
                value="2012"
              >
                2012
              </option>
              <option
                data-testid="option-year-2011"
                value="2011"
              >
                2011
              </option>
              <option
                data-testid="option-year-2010"
                value="2010"
              >
                2010
              </option>
              <option
                data-testid="option-year-2009"
                value="2009"
              >
                2009
              </option>
              <option
                data-testid="option-year-2008"
                value="2008"
              >
                2008
              </option>
              <option
                data-testid="option-year-2007"
                value="2007"
              >
                2007
              </option>
              <option
                data-testid="option-year-2006"
                value="2006"
              >
                2006
              </option>
            </select>
          </p>
          <button
            class="field btn "
            data-testid="submit-button"
            id="newBatchSubmit"
            type="submit"
          >
            <span
              class="btn__inner"
            >
              Submit
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
